### PR TITLE
Workflow: add third party calls

### PR DESCRIPTION
**Pull request by CahidArda**
additionally, this change includes:
- HTTPMethods type
- splitting the WorkflowContext.parseRequest method into two methods in workflow-parser.ts: validateRequest and parseRequest
- adding handleCallReturn method to workflow-requests.ts file to handle third party call result requests
- adding LazyCallStep to handle third party calls
- removing WorkflowContext.createContext and doing the same work in serve.ts

### Change in the serve method

As a result of the changes above, serve method changed from:
```ts
  return async (request: TRequest) => {
    const { workflowContext, isFirstInvocation } =
      await WorkflowContext.createContext<TInitialPayload>(request, client, initialPayloadParser);

    const result = isFirstInvocation
      ? await triggerFirstInvocation(workflowContext)
      : await triggerRouteFunction({
          onStep: async () => routeFunction(workflowContext),
          onCleanup: async () => triggerWorkflowDelete(workflowContext),
        });

    if (result.isErr()) {
      throw result.error;
    }

    // Returns a Response with `workflowId` at the end of each step.
    return onStepFinish(workflowContext.workflowId);
  };
```
 to:
```ts
  return async (request: TRequest) => {
    const callReturnCheck = await handleCallReturn(request, client);

    if (callReturnCheck.isErr()) {
      throw callReturnCheck.error;
    } else if (callReturnCheck.value === "continue-workflow") {
      const { isFirstInvocation, workflowId } = validateRequest(request);
      const { initialPayload, steps } = await parseRequest(request, isFirstInvocation);

      const workflowContext = new WorkflowContext<TInitialPayload>({
        client,
        workflowId,
        initialPayload: initialPayloadParser(initialPayload),
        steps,
        url: request.url,
      });

      const result = isFirstInvocation
        ? await triggerFirstInvocation(workflowContext)
        : await triggerRouteFunction({
            onStep: async () => routeFunction(workflowContext),
            onCleanup: async () => triggerWorkflowDelete(workflowContext),
          });

      if (result.isErr()) {
        throw result.error;
      }

      // Returns a Response with `workflowId` at the end of each step.
      return onStepFinish(workflowContext.workflowId);
    }

    // response to QStash in call cases
    return onStepFinish("no-workflow-id");
  };
```
I think the new version is clearer in terms of which method does what. In the older version, logic was hidden behind `createWorkflow`.

### CODE REVIEW

File: src/client/workflow/context.ts [hunk 1]
```diff
@@ -178,39 +127,41 @@ export class WorkflowContext<TInitialPayload = unknown> {
...
-      url: request.url,
-    });
+    if (step?.callHeaders && step.callMethod) {
+      const forwardedHeaders = Object.fromEntries(
+        Object.entries(step.callHeaders).map(([header, value]) => [
+          `Upstash-Forward-${header}`,
+          value,
+        ])
+      );
+      baseHeaders = {
```
Comment (ogzhanolguncu) [inline src/client/workflow/context.ts:144]: Why do we reassign? Can't we just mutate the object?

File: src/client/workflow/serve.ts [hunk 2]
```diff
@@ -74,21 +76,38 @@ export const serve = <
    * @returns A promise that resolves to a response.
    */
   return async (request: TRequest) => {
-    const { workflowContext, isFirstInvocation } =
-      await WorkflowContext.createContext<TInitialPayload>(request, client, initialPayloadParser);
+    const callReturnCheck = await handleCallReturn(request, client);
```
Comment (ogzhanolguncu) [inline src/client/workflow/serve.ts:79]: Let's document this and make intentions clearer. Why'd we be handlingCallReturn? If those are specifically for 3rd party calls maybe we can come up with a better name?

Comment (CahidArda) [reply src/client/workflow/serve.ts]: renamed handleCallReturn as handleThirdPartyCallResult and explained how third party calls work in handleThirdPartyCallResult docs

File: src/client/workflow/auto-executor.ts [hunk 3]
```diff
@@ -292,11 +256,14 @@ export class AutoExecutor {
...
         return {
-          headers: this.context.getHeaders("false"),
-          method: "POST",
-          body: JSON.stringify(singleStep),
-          url: this.context.url,
+          headers: this.context.getHeaders("false", singleStep, contentType),
+          // if the call parameters are set, they overwrite defaults
+          method: singleStep.callMethod ?? "POST",
+          body: singleStep.callBody ?? singleStep,
+          url: singleStep.callUrl ?? this.context.url,
```
[the remark this answers is not in the document]
Comment (CahidArda) [reply src/client/workflow/auto-executor.ts]: split this into two and explained what's going on with comments

File: src/client/workflow/context.ts (hunk 1 above)
Comment (CahidArda) [reply src/client/workflow/context.ts]: I removed the reassign. I moved the getHeaders from public API to workflow-requests.ts too. [1ef8619](https://github.com/upstash/qstash-js/pull/132/commits/1ef8619a61926ac6108b2c44653638ae219908c0)

**Review by ogzhanolguncu — APPROVED** [review]

**Pull request merged**